src/DominantJuice.sol

```diff
@@ -0,0 +1,396 @@
+        UD60x18 sumOfOrderWeights;
+        // Memory array of normalized order weights (w-1s)
+        uint256[] memory orderWeightNormalized = new uint[](numPledgers);
+        UD60x18 sumOfAmountsXNormalizedOrderWeights;
+        // uint256 donationWeight;
+        // uint256 donationWeightNormalized;
+        // Represents the combined order and pledge amount weighting.
+        UD60x18 overallPledgeWeight;
+        emit Test("totalAmountPledged", totalAmountPledged);
+        // First, get the sum of pledge order weights to use in the normalizing loop
+        for (uint32 i = 0; i < numPledgers; i++) {
+            // Add 1 to i because rateOfDecay is being raised to the pledge ordereth power
+            sumOfOrderWeights = add(sumOfOrderWeights, rateOfDecay.powu(i + 1));
+        }
+
+        // Refund Bonus calculation prep work:
+        for (uint32 j = 0; j < numPledgers; j++) {
+            // Earlier pledges are given more weight according to the exponential decay rate.
+            // Add 1 to j because rateOfDecay is being raised to the ordereth power.
+            orderWeight = rateOfDecay.powu(j + 1);
+            // Order weights are normalized to compare each pledge to the total pledging data.
+            // Save in memory array for use in next for loop.
+            orderWeightNormalized[j] = unwrap(orderWeight / sumOfOrderWeights);
+            // Get the payment amount for each order
+            pledgeAmount_orderLoop = wrap(orderToPledgerToAmount[j][pledgers[j]]);
+            // Multiply payments by their respective normalized weight. Sum all products by
+            // looping (=sum(x*w_1)). Use in next for loop
+            sumOfAmountsXNormalizedOrderWeights =
+                add(sumOfAmountsXNormalizedOrderWeights, mul(pledgeAmount_orderLoop, wrap(orderWeightNormalized[j])));
+        }
+
+        // Next, give more weight to larger contribution amounts by multiplying each pledge amount with its
+        // orderWeightNormalized. Looping logic is for the possibility of repeat pledgers
+        for (uint32 k = 0; k < numPledgers; k++) {
+            if (pledgers[k] == _pledger) {
+                pledgeAmount_amountLoop = wrap(orderToPledgerToAmount[k][_pledger]);
+                overallPledgeWeight =
+                    mul(pledgeAmount_amountLoop, wrap(orderWeightNormalized[k])) / sumOfAmountsXNormalizedOrderWeights;
+                // Multiply totalRefundBonus by overallWeight to get the individual refund. Loop again.
+                // If pledger's address is found in the pledger array again, add to individual total.
+                individualRefundBonus = add(individualRefundBonus, mul(overallPledgeWeight, wrap(totalRefundBonus)));
+            }
+        }
+        return unwrap(individualRefundBonus);
+    }
+
+    function getBalance() public view returns (uint256) {
+        return address(this).balance;
+    }
+
+    function getCycleFundingStatus() public view returns (uint256, uint256, bool, bool) {
```

Comment:
this would be better described as a struct. anytime more than 1 or 2 return values are used, it's typically better to use a struct. it doesn't look like it's part of the interface

Reply:
And just for my education again, what's the reasoning here? Is it an ABI thing?

And you mean returning a struct right?

Reply:
this is a code comprehension and maintainability thing (almost all languages). You can see this when passing variables to a function as well. Too many variables in a row in the parameter list (passing in or returning) makes it hard to maintain, especially when there are multiple bool. If you can force callers to pass by name (eg keyword args in Python), then the code can be self-commenting. But imagine calling a function like this:

`doit("something", "something else here", 1, 10, 12.2, true, false, true)`

You can't understand it when you're reading the code. This same can be said for returns

`return "something", "something else", true, false`

And then the caller has to deal with multiple variables on the return:

`name, description, is_done, is_working = doit()`

It's a lot more descriptive to pass around structs for these cases.

Reply:
Understood, thank you.